### ISSUE TITLE
Threadsafe?

### ISSUE THREAD

**Issue by snadrus** [go-critic/go-critic#936]
Is it possible to make & grow a version of this which could guarantee thread safety of memory accesses. 

V1: 
Use of go keyword:
  - only gets non-pointer args 
  - inline funcs: context which is not further used in parent or concurrent Go.
  - Cannot be reached from a "func init" or global "var X = func(){}" 

Globals:
 - Are not modified after init except with a mutex that's locked & reused for reads. 

V2: To enable pointer passing, this could get complicated like the Rust borrow checker.

**Comment by cristaloleg** [comment]
> only gets non-pointer args

but there is no problem with a given code:

```
type S struct {
	mu sync.Mutex
	v  int64
}

func main() {
	s := &S{}
	go foo(s)
}

func foo(s *S) {
	s.mu.Lock()
	s.v++
	s.mu.Unlock()
}
```

**Comment by snadrus** [comment]
There are many safe cases not covered in my V1 proposal, but allowing no false positives makes it a valuable tool while it grows.

V1 should also warn on sync, atomic, CGo & unsafe being unimplemented. The former two eventually will get implemented to some extent.

V2 should be a "single owner" prover which would enable your example to pass. This is the core of my question: could such a rich traversal be reasonable in this architecture?

Then a future version could use that for a checker to validate a variable going from single-owner 'mode' into mutex-protected 'mode'.

**Comment by cristaloleg** [comment]
Can you clarify:
> should also warn on sync, atomic <...> being unimplemented

What do you mean?

**Comment by snadrus** [comment]
The whole idea is to either prove safety or error with "could not prove whether this program is threadsafe".

That 'warn' talk was my debating adding to that error some info that sync and atomic checking are unimplemented, but I'm content either way.
